server/methods/core/groups.js

```diff
@@ -0,0 +1,207 @@
+
+  /**
+   * group/updateGroup
+   * @summary updates a permission group for a shop
+   * updates either the name of the permission group or it list of permissions
+   * It also goes into affected user data to modify both the groupName (using Accounts schema)
+   * and group permissions (using "accounts/removeUserPermissions")
+   * @param {Object} group - current data of the group to be updated
+   * @param {String} group.name - name of the group
+   * @param {Array} group.permissions - permissions of the grop
+   * @param {Object} newGroupData - updated group info (similar to current group data)
+   * @param {String} shopId - id of the shop the group belongs to
+   * @return {Object} - object.status of 200 on success or Error object on failure
+   */
+  "group/updateGroup": function (group, newGroupData, shopId) {
+    check(group, Object);
+    check(group.name, String);
+    check(group.permissions, [String]);
+
+    check(newGroupData, Object);
+    check(newGroupData.name, String);
+    check(newGroupData.permissions, [String]);
+    check(shopId, String);
+
+    // must have needed permissions.. .TODO: Review who can create a group permission for a shop
+    if (!Roles.userIsInRole(Meteor.userId(), "admin", shopId)) {
+      throw new Meteor.Error(403, "Access Denied");
+    }
+
+    const permissionsChanged = !_.isEqual(group.permissions, newGroupData.permissions);
+    const updateQuery = { "groups.$": newGroupData };
+
+    Shops.update({ _id: shopId, groups: group }, { $set: updateQuery });
+
+    const matchQuery = { $elemMatch: { shopId: shopId, names: group.name } };
+    const affectedUsers = Accounts.find({ groups: matchQuery }).fetch();
+    let error;
+
+    if (permissionsChanged) {
+      error = updateUsersPermissions(affectedUsers, group, newGroupData, shopId);
+    }
+
+    // update group names on user account
+    error = updateUsersGroupName(affectedUsers, shopId);
+
+    if (!error) {
+      return { status: 200 };
+    }
+    Logger.error(error);
+    throw new Meteor.Error(500, "Update not successful");
+  },
+  /**
+   * group/addUser
+   * @summary adds a user to a permission group for a shop
+   * It updates the user's list of permissions/roles with the defined the list defined for the group
+   * It uses the local *updateUsers-* by passing one user as an array
+   * @param {String} userId - current data of the group to be updated
+   * @param {Object} groupData - info about the group
+   * @param {String} shopId - permissions of the group
+   * @return {Object} - object.status of 200 on success or Error object on failure
+   */
+  "group/addUser": function (userId, groupData, shopId) {
+    check(userId, String);
+    check(groupData, Object);
+    check(groupData.name, String);
+    check(groupData.permissions, [String]);
+    check(shopId, String);
+
+    const user = Accounts.findOne({ _id: userId });
+
+    try {
+      Meteor.call("accounts/addUserPermissions", user.userId, groupData.permissions, shopId);
```

Comment:
One thing I've been thinking about is that it might be easier to do group assignment if any user can only have one group per shop for now. You could then add/remove the difference in permissions between users old group and users new group and just add the group name. It would negate the need to do the hard determining which group(s) a user is in by their permissions.